ticker/src/main/java/com/robinhood/ticker/TickerCharacterList.java

```diff
@@ -0,0 +1,114 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.robinhood.ticker;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This is the primary class that Ticker uses to determine how to animate from one character
+ * to another. The provided string dictates what characters will appear between
+ * the start and end characters.
+ *
+ * <p>For example, given the string "abcde", if the view wants to animate from 'd' to 'b',
+ * it will know that it has to go from 'd' to 'c' to 'b', and these are the characters
+ * that show up during the animation scroll.
+ *
+ * @author Jin Cao, Robinhood
+ */
+public class TickerCharacterList {
+    private final int numOriginalCharacters;
+    // The saved character list will always be of the format: EMPTY, list, list
+    private final char[] characterList;
+    // A minor optimization so that we can cache the indices of each character.
+    private final Map<Character, Integer> characterIndicesMap;
+
+    public TickerCharacterList(String characterList) {
+        if (characterList.contains(Character.toString(TickerUtils.EMPTY_CHAR))) {
+            throw new IllegalArgumentException(
+                    "You don't need to include TickerUtils.EMPTY_CHAR in the character list.");
+        }
+
+        final char[] charsArray = characterList.toCharArray();
+        final int length = charsArray.length;
+        this.numOriginalCharacters = length;
+
+        characterIndicesMap = new HashMap<>(length);
+        for (int i = 0; i < length; i++) {
+            characterIndicesMap.put(charsArray[i], i);
+        }
+
+        this.characterList = new char[length * 2 + 1];
+        this.characterList[0] = TickerUtils.EMPTY_CHAR;
+        for (int i = 0; i < length; i++) {
+            this.characterList[1 + i] = charsArray[i];
+            this.characterList[1 + length + i] = charsArray[i];
+        }
+    }
+
+    /**
+     * @param start the character that we want to animate from
+     * @param end the character that we want to animate to
+     * @return a valid {@link AnimationCharacterIndices}, or null if the inputs are not supported.
+     */
+    AnimationCharacterIndices getCharacterIndices(char start, char end) {
+        int startIndex = getIndexOfChar(start);
+        int endIndex = getIndexOfChar(end);
+        if (startIndex < 0 || endIndex < 0) {
+            return null;
+        }
+
+        if (start != TickerUtils.EMPTY_CHAR && end != TickerUtils.EMPTY_CHAR &&
+                endIndex < startIndex) {
+            final int nonWrapDistance = startIndex - endIndex;
+            final int wrapDistance = numOriginalCharacters - startIndex + endIndex;
+            if (wrapDistance < nonWrapDistance) {
+                endIndex = numOriginalCharacters + endIndex;
```

Comment:
yeah, i'm still not 100% sure about the second list thing -

say we wraparound from 9 -> 3, then we go from 3 -> 7, 7 -> 9

at this point, are we displaying the 9 from the second list? if so, i think if we try to go from 9 -> 3 again, won't it always go down instead of wrapping around?

Reply:
see explanation above, the indices for animation will always use the first list, the second list is entirely for drawing the animation.